src/cluster_asm.c

```diff
@@ -339,3 +424,616 @@ void clusterMigrationCommand(client *c) {
         addReplyError(c, "unknown argument");
     }
 }
+
+/* Sends an AUTH command to the source node using the internal secret.
+ * Returns an error string if the command fails, or NULL on success. */
+char *asmSendInternalAuth(connection *conn) {
+    size_t len = -1;
+    const char *internal_secret = clusterGetSecret(&len);
+    serverAssert(len > 0);
+
+    sds secret = sdsnewlen(internal_secret, len);
+    char *err = sendCommand(conn, "AUTH", "internal connection", secret, NULL);
+    sdsfree(secret);
+    return err;
+}
+
+void asmRdbChannelSyncWithSource(connection *conn) {
+    asmTask *task = connGetPrivateData(conn);
+    char *err = NULL;
+
+    if (task->rdb_channel_state == ASM_CONNECTING) {
+        connSetReadHandler(conn, asmRdbChannelSyncWithSource);
+        connSetWriteHandler(conn, NULL);
+
+        /* Send AUTH command to source node using internal auth */
+        err = asmSendInternalAuth(conn);
+        if (err) goto write_error;
+        task->rdb_channel_state = ASM_AUTH_REPLY;
+        return;
+    }
+
+    if (task->rdb_channel_state == ASM_AUTH_REPLY) {
+        err = receiveSynchronousResponse(conn);
+        /* The source node did not reply */
+        if (err == NULL) goto no_response_error;
+
+        /* Check `+OK` reply */
+        if (!strcmp(err, "+OK")) {
+            sdsfree(err);
+            err = NULL;
+            task->rdb_channel_state = ASM_SEND_RDBCHANNEL;
+            serverLog(LL_NOTICE, "Source node replied to AUTH command, slots rdb channel sync can continue...");
+        } else {
+            serverLog(LL_WARNING, "Error reply to AUTH from source: '%s'", err);
+            sdsfree(err);
+            goto error;
+        }
+    }
+
+    if (task->rdb_channel_state == ASM_SEND_RDBCHANNEL) {
+        char cid[LONG_STR_SIZE];
+        ull2string(cid, sizeof(cid), task->main_channel_id);
+
+        err = sendCommand(conn, "CLUSTER", "SYNCSLOTS", "RDBCHANNEL", cid, NULL);
+        if (err) goto write_error;
+        task->rdb_channel_state = ASM_RDBCHANNEL_REPLY;
+        return;
+    }
+
+    if (task->rdb_channel_state == ASM_RDBCHANNEL_REPLY) {
+        err = receiveSynchronousResponse(conn);
+        /* The destination node did not reply */
+        if (err == NULL) goto no_response_error;
+
+        /* Check `+SLOTSSNAPSHOT` reply */
+        if (!strncmp(err, "+SLOTSSNAPSHOT", strlen("+SLOTSSNAPSHOT"))) {
+            task->state = ASM_BUFFER_STREAM;
+            /* TODO: buffer pending commands stream */
+            // connSetReadHandler(task->main_channel_con, mainChannelBufferStream);
+
+            task->rdb_channel_state = ASM_RDBCHANNEL_TRANSFER;
+            client *c = createClient(conn);
+            c->flags |= CLIENT_MASTER;
+            c->querybuf = sdsempty();
+            c->authenticated = 1;
+            c->user = NULL;
+            c->task = task;
+            connSetPrivateData(conn, c);
+            connSetReadHandler(conn, readQueryFromClient);
```

Comment:
```suggestion
```

I think `createClient(conn)` does these two calls.

Reply:
yes, you are right